docs_nnx/guides/transforms.md

```diff
@@ -0,0 +1,340 @@
+  weights.count += 1
+  return x @ weights.kernel + weights.bias
+
+y = nnx.vmap(stateful_vector_dot, in_axes=0, out_axes=1)(weights, x)
+y = nnx.vmap(stateful_vector_dot, in_axes=0, out_axes=1)(weights, x)
+
+weights.count
+```
+
+Here we ran `stateful_vector_dot` twice to verify that the `count` attribute was correctly updated. Because Weights is vectorized, `count` was initialized as an `arange(10)`, and all of its elements were incremented by 2 inside the transformation. The most important part is that updates were propagated to the original `Weights` object outside the transformation. Nice!
+
++++
+
+### Graph updates propagation
+JAX models inputs to transformations as trees, Flax NNX models inputs as graphs to allow for sharing references. However, to express most of Python's object model Flax NNX's state propagation machinery can track arbitrary updates to the objects as long as they're local (updates to globals inside transforms are not supported). This means that you can modify graph structure as needed, including updating existing attributes, adding/deleting attributes, swapping attributes, sharing (new) references between objects, sharing Variables between objects, etc. The sky is the limit!
+
+The following example demonstrates performing some arbitrary updates to the `Weights` object inside `nnx.vmap` and verifying that the updates are correctly propagated to the original `Weights` object outside the transformation.
+
+```{code-cell} ipython3
+class Count(nnx.Variable): pass
+
+class Weights(nnx.Module):
+  def __init__(self, kernel: jax.Array, bias: jax.Array, count: jax.Array):
+    self.kernel, self.bias = nnx.Param(kernel), nnx.Param(bias)
+    self.count = Count(count)
+
+weights = Weights(
+  kernel=random.uniform(random.key(0), (10, 2, 3)),
+  bias=jnp.zeros((10, 3)),
+  count=jnp.arange(10),
+)
+x = jax.random.normal(random.key(1), (10, 2))
+
+def crazy_vector_dot(weights: Weights, x: jax.Array):
+  assert weights.kernel.ndim == 2, 'Batch dimensions not allowed'
+  assert x.ndim == 1, 'Batch dimensions not allowed'
+  weights.count += 1
+  y = x @ weights.kernel + weights.bias
+  weights.some_property = ['a', 2, False] # add attribute
+  del weights.bias # delete attribute
+  weights.new_param = weights.kernel # share reference
+  return y
+
+y = nnx.vmap(crazy_vector_dot, in_axes=0, out_axes=1)(weights, x)
+
+nnx.display(weights)
+```
+
+> With great power comes great responsibility.
+> <br> \- Uncle Ben
+
+While this feature is very powerful, it must be used with care as it can clash with JAX's underlying assumptions for certain transformations. For example, `jit` expects the structure of the inputs to be stable in order to cache the compiled function, changing the graph structure inside a `nnx.jit`-ed function cause continuous recompilations and performance degradation, `scan` on the other hand only allows a fixed `carry` structure, so adding/removing substates declared as carry will cause an error.
```

Comment:
nit:
> For example, `jit` expects the structure of the inputs to be stable in order to cache the compiled function, changing the graph structure inside a `nnx.jit`-ed function cause continuous recompilations and performance degradation, `scan` on the other hand only allows a fixed `carry` structure, so adding/removing substates declared as carry will cause an error.

For example, `jit` expects the structure of the inputs to be stable in order to cache the compiled function, so changing the graph structure inside a `nnx.jit`-ed function cause continuous recompilations and performance degradation. On the other hand, `scan` only allows a fixed `carry` structure, so adding/removing substates declared as carry will cause an error.